### PR TITLE
fix(dashboard list): do not show favorite star for anonymous users  #18210

### LINKED ISSUE #18210: Unexpected error: ListViewError: Invalid filter config, id is not present in columns
Prevent 'undefined' error when viewing the dashboard (or charts) listview in an anonymous mode.

#### How to reproduce the bug

Reproduce the error as follows:

Allow anonymous access by setting `PUBLIC_ROLE_LIKE = "Gamma"`
Without logging in, click on dashboards or charts. This error displays:
_Unexpected error:ListViewError: Invalid filter config, id is not present in columns_

### Expected results
Dashboards/charts must be visible without any issues/errors.

### Actual result
This issue observed when we make upgrade from v1.3 to v1.4

#### Screenshots

image

### Environment

(please complete the following information):

- browser type and version: `Chrome Version 97.0.4692.71`
- superset version: `1.4`
- python version: `Python 3.7.5`
- node.js version: `v8.10.0`

### Checklist

- [x] I have checked the superset logs for python stacktraces and included it here as text if there are any.
- [x] I have reproduced the issue with at least the latest released version of superset.
- [x] I have checked the issue tracker for the same issue and I haven't found one similar.

### Additional context

Browser console log:

_**ListViewError: Invalid filter config, id is not present in columns
    at 87c5db28d6d4ce263182.chunk.js:932:15
    at Array.forEach (<anonymous>)
    at ListView (87c5db28d6d4ce263182.chunk.js:930:13)
    at oh (vendors.866d9853ec9ca701f3b8.entry.js:54972:146)
    at Rj (vendors.866d9853ec9ca701f3b8.entry.js:55080:496)
    at Qj (vendors.866d9853ec9ca701f3b8.entry.js:55065:199)
    at Kj (vendors.866d9853ec9ca701f3b8.entry.js:55065:128)
    at yj (vendors.866d9853ec9ca701f3b8.entry.js:55058:172)
    at vendors.866d9853ec9ca701f3b8.entry.js:54942:115
    at exports.unstable_runWithPriority (vendors.866d9853ec9ca701f3b8.entry.js:55173:467)**_
    
    
image

### PR DESCRIPTION
**Pull request by dudasaron**
### SUMMARY
A previous commit removed the `FaveStar` component and the `id` column with it from the dashboards list view for anonymous users. This broke the `CERTIFIED` filter because it also depends on the `id` column.
This PR re-adds the id column, but removes the favorite star and hides the column if the user is not logged in. Also removing the stars from the `DashboardCard` component if not applicable.

#### Before ####
Anonymous user, list/card view:
image

Logged in user, dashboard (*for reference, not supposed to change*):
image

Logged in user, list view (*for reference, not supposed to change*):
image

Logged in user, card view (*for reference, not supposed to change*):
image

#### After ####
Anonymous user, list view:
image

Anonymous user, card view:
image

Logged in user, dashboard (*not supposed to change visually, but screenshot is included because some components are modified*):
image

Logged in user, list view (*not supposed to change visually, but screenshot is included because some components are modified*):
image

Logged in user, card view (*not supposed to change visually, but screenshot is included because some components are modified*):
image

### TESTING INSTRUCTIONS
Automated tests added for both logged in/anonymous cases for table and card view.

For manual testing add 
```python
PUBLIC_ROLE_LIKE = "Gamma"
```
to `docker/pythonpath_dev/superset_config.py` and run `docker compose up`.

Log in as admin and add the `all datasource access on all_datasource_access` and `all database access on all_database_access` permissions to the public role.

After logging out, navigate to [http://localhost:3000/superset/dashboards/list](http://localhost:3000/superset/dashboards/list) and the dashboard list should open without an error and the CERTIFIED filter should work and the favorite star should not be present on the list view nor the table view. When logging in, the favorite star should be there.

### ADDITIONAL INFORMATION
- [x] Has associated issue: Fixes #18210
- [ ] Required feature flags:
- [x] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/src/views/CRUD/dashboard/DashboardCard.tsx [hunk 1]
```diff
@@ -44,7 +44,7 @@ interface DashboardCardProps {
   saveFavoriteStatus: (id: number, isStarred: boolean) => void;
   favoriteStatus: boolean;
   dashboardFilter?: string;
-  userId?: number;
+  userId?: string | number;
```
Comment (AAfghahi) [inline superset-frontend/src/views/CRUD/dashboard/DashboardCard.tsx:47]: Hey! I don't think that this should ever be a string? Could you give me more context?

File: superset-frontend/src/views/CRUD/dashboard/DashboardList.tsx [hunk 2]
```diff
@@ -232,27 +235,25 @@ function DashboardList(props: DashboardListProps) {
...
-            },
-          ]
-        : []),
+      {
+        Cell: ({
+          row: {
+            original: { id },
+          },
+        }: any) =>
+          userId ? (
```
Comment (AAfghahi) [inline superset-frontend/src/views/CRUD/dashboard/DashboardList.tsx:244]: Why this over userId && <Component>

**Review by AAfghahi — CHANGES_REQUESTED** [review]

File: superset-frontend/src/views/CRUD/dashboard/DashboardList.tsx (hunk 2 above)
Comment (dudasaron) [reply superset-frontend/src/views/CRUD/dashboard/DashboardList.tsx:244]: good point, I'll change it

File: superset-frontend/src/views/CRUD/dashboard/DashboardCard.tsx (hunk 1 above)
Comment (dudasaron) [reply superset-frontend/src/views/CRUD/dashboard/DashboardCard.tsx:47]: I found it strange too, but it comes from already existing code: `DashboardList.tsx:73` and I didn't want to dig deep into this for this PR

**Review by AAfghahi — APPROVED** [review]

File: superset-frontend/src/views/CRUD/dashboard/DashboardCard.tsx (hunk 1 above)
Comment (pkdotson) [reply superset-frontend/src/views/CRUD/dashboard/DashboardCard.tsx:47]: I think this should be number. You can take note of it for a follow up.

Comment (pkdotson) [reply superset-frontend/src/views/CRUD/dashboard/DashboardCard.tsx:47]: I think this should be number. You can take note of it for a follow up.

**Review by pkdotson — APPROVED** [review]
LGTM!

**Review by villebro — APPROVED** [review]
This is great work @dudasaron ! Thanks not only for fixing this, but also cleaning up the code + adding tests ❤️

**Comment by crazamahdi** [comment]
Hey I am facing same error....please let me know how to fix that issue....i read the whole conversation but did not understand what changes to be done in config.py file.....I wrote PUBLIC_ROLE_LIKE = "Gamma" in config.py file then relaunched it through superset init but still i face the same unexpected error..ListViewError: Invalid filter config, id is not present in columns...pls Help me out  @dudasaron or anyone who fixed this issue.

**Comment by dudasaron** [comment]
Hi @crazamahdi, I think you were referring to the reproduction steps of the bug. What version are you using? This fix is only merged into master, but it's not yet part of any released version. If you don't have the time to wait until it makes into the official release, you can create a fork of the latest release of superset (1.4.2 as of now) and cherry pick the merge commit on top of that.

**Comment by crazamahdi** [comment]
> Hi @crazamahdi, I think you were referring to the reproduction steps of the bug. What version are you using? This fix is only merged into master, but it's not yet part of any released version. If you don't have the time to wait until it makes into the official release, you can create a fork of the latest release of superset (1.4.2 as of now) and cherry pick the merge commit on top of that.

Hey thanks for your reply …I am using version 1.4.1…but I have superset version 1.4.1 on my local machine which works fine (running on http)…now when I installed the same version on my work machine(Aws instance Ubuntu 20) it throws me this unexpected error ListViewError….(while running  on HTTPS)…I don’t understand how same version of superset is working on my local machine and not on my work machine….is it something related to HTTP and HTTPS? If so what changes should I do in config file to run it on https. For me I am not even able to login through admin…after entering credentials this unexpected error screen shows up

I tried installing, deleting and reinstalling various versions of superset on my work machine but everytime it throws some unexpected error….please help me out with this issue I tried resolving it for more than 4 days but no success yet@dudasaron

**Comment by chenshaojie** [comment]
> > Hi @crazamahdi, I think you were referring to the reproduction steps of the bug. What version are you using? This fix is only merged into master, but it's not yet part of any released version. If you don't have the time to wait until it makes into the official release, you can create a fork of the latest release of superset (1.4.2 as of now) and cherry pick the merge commit on top of that.
> 
> Hey thanks for your reply …I am using version 1.4.1…but I have superset version 1.4.1 on my local machine which works fine (running on http)…now when I installed the same version on my work machine(Aws instance Ubuntu 20) it throws me this unexpected error ListViewError….(while running on HTTPS)…I don’t understand how same version of superset is working on my local machine and not on my work machine….is it something related to HTTP and HTTPS? If so what changes should I do in config file to run it on https. For me I am not even able to login through admin…after entering credentials this unexpected error screen shows up
> 
> I tried installing, deleting and reinstalling various versions of superset on my work machine but everytime it throws some unexpected error….please help me out with this issue I tried resolving it for more than 4 days but no success yet@dudasaron

Hello, I took the liberty of disturbing, I am using the same version of the program, but I use centos, and now I am encountering the same problem as you, do you have a solution now? I found myself that even if I logged in, I would always stay on the first page of the previous page described by the author, similar to the refresh operation, but I myself used wechat to browse the web page, although there would be an operation described by the author, I could still log in and show two or three pictures of the previous page. Thank you very much. 
Sorry, my English is not good, so use the translator for conversations, thank you very much for answering my questions

**Comment by crazamahdi** [comment]
Hey @chenshaojie I am still stuck at the same issue…hoping they release the fixed version soon. Got to know that issue is from their end and the fix is not yet officially released but committed to master branch. U can wait for release or u can try from GitHub directly. (If GitHub one works pls let me know too)
Thanks

**Comment by crazamahdi** [comment]
Hey @chenshaojie I finally managed to solved that error. Solve the login error by commenting the following lines in config.py file
# Comment settings as below:
#SESSION_COOKIE_SAMESITE = "None"
#SESSION_COOKIE_SECURE = True

someone helped me with the solution on stack overflow.
and it worked well for me.

**Comment by chenshaojie** [comment]
thank you very much !૧(●´৺`●)૭ I will try it 😘



---Original---
From: "Raza Mahdi ***@***.***&gt;
Date: Sun, Apr 24, 2022 11:05 AM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [apache/superset] fix(dashboard list): do not show favorite starfor anonymous users  #18210 (PR #19409)




 
Hey @chenshaojie I finally managed to solved that error. Solve the login error by commenting the following lines in config.py file
 
Comment settings as below:
 
#SESSION_COOKIE_SAMESITE = "None"
 #SESSION_COOKIE_SECURE = True
 
someone helped me with the solution on stack overflow.
 and it worked well for me.

**Pull request merged**